python/ray/tune/ray_trial_executor.py

```diff
@@ -380,33 +453,68 @@ def _start_trial(self, trial, checkpoint=None, runner=None,
             self._train(trial)
         return True
 
-    def _stop_trial(self, trial, error=False, error_msg=None):
+    def _stop_trial(self, trial, error=False, error_msg=None, free=True):
         """Stops this trial.
 
         Stops this trial, releasing all allocating resources. If stopping the
         trial fails, the run will be marked as terminated in error, but no
         exception will be thrown.
 
+        If the trial should be paused (``pause=True``), we do not remove
+        its placement group (or a surrogate placement group).
+
         Args:
             error (bool): Whether to mark this trial as terminated in error.
             error_msg (str): Optional error message.
+            free (bool): Whether this trials resources (placement groups)
+                should be free'd.
         """
         self.set_status(trial, Trial.ERROR if error else Trial.TERMINATED)
+        self._trial_just_finished = True
         trial.set_location(Location())
 
         try:
             trial.write_error_log(error_msg)
             if hasattr(trial, "runner") and trial.runner:
                 if (not error and self._reuse_actors
-                        and self._cached_actor is None):
+                        and self._cached_actor_pg[0] is None):
                     logger.debug("Reusing actor for %s", trial.runner)
-                    self._cached_actor = trial.runner
+                    # Move PG into cache (disassociate from trial)
+                    pg = self._pg_manager.cache_trial_pg(
+                        trial, replace_pending=free)
+                    if pg:  # Always true if replace_pending=False
+                        self._cached_actor_pg = (trial.runner, pg)
+                        should_destroy = False
+                    else:
+                        logger.debug(
+                            "Could not cache of trial {trial} actor for "
+                            "reuse, as there are no pending trials "
+                            "requiring its resources.")
+                        should_destroy = True
                 else:
+                    should_destroy = True
+
+                if should_destroy:
                     logger.debug("Trial %s: Destroying actor.", trial)
-                    pg = self._pg_manager.clean_trial_placement_group(trial)
+
+                    # Try to return the placement group for other trials to use
+                    if not self._pg_manager.return_pg(
+                            trial, replace_pending=free):
```

Comment:
question: if you are planning to free the placement group, does that mean you're planning to delete the placement group?

Reply:
Hmm, why is it that when you return a placement group for use by other actors that you'll need to destroy another equal placement group? Is it because there may be a placement group that is pending?

Would it be possible to just return it to a pool of placement groups, and do placement group clean up at a different frequency?

Reply:
(maybe this question doesn't even make sense!)

Reply:
Done (see more elaborate comment below)